src/packages/hex.gleam

```diff
@@ -1,55 +1,180 @@
 import gleam/io
 import gleam/json
-import gleam/http
+import gleam/http/request
+import gleam/http.{Get}
 import gleam/result
 import gleam/hackney
-import gleam/dynamic.{Decoder} as d
+import gleam/list
+import gleam/dynamic as d
+import time/parse.{parse_iso8601_to_gregorian_seconds}
+import gleam/pgo
+import gleam/string
+import gleam/int
+import packages/models/hex/package.{HexPackage, hex_package_decoder}
+import packages/models/hex/release.{hex_release_decoder}
 
 pub type Error {
   HackneyError(hackney.Error)
   JsonError(json.DecodeError)
+  DatabaseError(pgo.QueryError)
+  ItemNotInListError(Nil)
 }
 
-pub fn query() {
+pub fn query(db: pgo.Connection) {
+  try last_scanned = get_last_scanned(db)
+
+  try packages = query_all_packages([], 1, last_scanned)
+
+  // Get the packages into order bassed on when they have
+  // been updated with new data
+  let packages =
+    packages
+    |> list.sort(fn(a, b) {
+      int.compare(
+        parse.parse_iso8601_to_gregorian_seconds(a.updated_at),
+        parse.parse_iso8601_to_gregorian_seconds(b.updated_at),
+      )
+    })
+
+  // Only update last scanned at if packages were actually
+  // found that way we only ever use the time from
+  // Hex's API, the list is revsersed so that the last package
+  // is actually the first for easy with the case statement
+  case
+    packages
+    |> list.reverse
+  {
+    [last, ..] -> {
+      assert Ok(_) =
+        update_last_scanned(db, parse.parse_iso8601(last.updated_at))
+      Ok(Nil)
+    }
+    _ -> Ok(Nil)
+  }
+
+  packages
+  |> list.filter_map(fliter_map_packages)
+  // TODO (HarryET): Remove before merge of PR
+  |> io.debug
+  |> Ok
+}
+
+const last_scanned_query = "SELECT date_trunc('second', scanned_at) FROM previous_hex_api_scan LIMIT 1;"
+
+fn get_last_scanned(db: pgo.Connection) -> Result(Int, Error) {
+  let ints = d.tuple3(d.int, d.int, d.int)
+  let datetime_decoder = d.tuple2(ints, ints)
+
+  try response =
+    pgo.execute(last_scanned_query, db, [], d.element(0, datetime_decoder))
+    |> result.map_error(DatabaseError)
+
+  case response.rows {
+    [first, ..] -> Ok(parse.to_gregorian_seconds(first))
+    _ -> Ok(0)
+  }
+}
+
+fn update_last_scanned(
+  db: pgo.Connection,
+  date: parse.LocalDateTime,
+) -> Result(pgo.Returned(d.Dynamic), Error) {
+  // It is done this way as I could not figure out how to
+  // inject a time as a string with the built in $1 system but
+  // this is safe as the value is constructed from a known datatype
+  pgo.execute(
+    "UPDATE previous_hex_api_scan SET scanned_at = '"
+    |> string.append(parse.to_pg_time(date))
+    |> string.append("';"),
+    db,
+    [],
+    d.dynamic,
+  )
+  |> result.map_error(DatabaseError)
+}
+
+pub fn fliter_map_packages(package: HexPackage) -> Result(HexPackage, Error) {
```

Comment:
It's not clear to me what this function does and it has no documentation. Could we add this on to explain what the intent is? There may also be a clearer name